### PR TITLE
osd/PG: two cleanups

### PR DESCRIPTION
**Pull request by xiexingguo**
"all_info" includes peers from up and acting sets too, thus
 deduplication is still needed here when calculating want_acting set.

### CODE REVIEW

**Comment by xiexingguo** [comment]
@jdurgin Since you are there, mind taking a look?

File: src/osd/PG.cc [hunk 1]
```diff
@@ -1168,7 +1168,8 @@ void PG::calc_ec_acting(
     if (acting.size() > (unsigned)i && acting[i] != CRUSH_ITEM_NONE &&
 	!all_info.find(pg_shard_t(acting[i], shard_id_t(i)))->second.is_incomplete() &&
 	all_info.find(pg_shard_t(acting[i], shard_id_t(i)))->second.last_update >=
-	auth_log_shard->second.log_tail) {
+	auth_log_shard->second.log_tail &&
+        std::find(want.begin(), want.end(), acting[i]) == want.end()) {
```
Comment (tchaikov) [inline src/osd/PG.cc]: there is chance that two shards are co-located in the same OSD. so i don't think we should dedup here.

File: src/osd/PG.cc [hunk 2]
```diff
@@ -1177,6 +1178,10 @@ void PG::calc_ec_acting(
 	   j != all_info_by_shard[shard_id_t(i)].end();
 	   ++j) {
 	assert(j->shard == i);
+        if (j->osd == CRUSH_ITEM_NONE ||
```
Comment (tchaikov) [inline src/osd/PG.cc]: it's fine if `j->osd == CRUSH_ITEM_NONE`.

File: src/osd/PG.cc [hunk 3]
```diff
@@ -1302,12 +1302,12 @@ void PG::calc_replicated_acting(
     const pg_info_t &cur_info = all_info.find(acting_cand)->second;
     if (cur_info.is_incomplete() ||
 	cur_info.last_update < primary->second.log_tail) {
-      ss << " shard " << acting_cand << " (stray) REJECTED "
+      ss << " shard " << acting_cand << " (acting) REJECTED "
```
Comment (tchaikov) [inline src/osd/PG.cc]: lgtm.

File: src/osd/PG.cc (hunk 1 above)
Comment (xiexingguo) [inline src/osd/PG.cc]: Thanks for pointing out this for me, kefu. I confess that I don't understand the behaviour of EC-PGs well:-(

**Comment by yuriw** [comment]
http://pulpito.ceph.com/yuriw-2017-08-23_23:03:38-rados-wip-yuri-testing4_2017_8_24-distro-basic-smithi/

http://pulpito.ceph.com/yuriw-2017-08-24_14:50:56-rados-wip-yuri-testing4_2017_8_24-distro-basic-smithi/

**Pull request merged**